include/range/v3/view/cycle.hpp

```diff
@@ -0,0 +1,259 @@
+                    }
+                }
+
+                bool equal(cursor const &pos) const
+                {
+                    return rng_ == pos.rng_ && it_ == pos.it_ && pos_ == pos.pos_;
+                }
+
+                CONCEPT_REQUIRES(BidirectionalRange<Rng>{})
+                void prev()
+                {
+                    --pos_;
+                    if (it_ == begin(rng_->rng_))
+                    {
+                        it_ = end(rng_->rng_);
+                    }
+                    --it_;
+                }
+
+                CONCEPT_REQUIRES(RandomAccessRange<Rng>())
+                void advance(difference_type n)
+                {
+                    if (n == difference_type{0})
+                    {
+                        return;
+                    }
+
+                    range_difference_t<Rng> distance = ranges::distance(rng_->rng_);
+                    if (distance == difference_type{0})
+                    {
+                        return;
+                    }
+
+                    pos_ += n;
+                    const auto begin = ranges::begin(rng_->rng_);
+                    const auto end = ranges::end(rng_->rng_);
+
+                    const difference_type pos = it_ - begin;
+                    const auto rel_distance = n >= 0 ? distance - pos : pos;
+
+                    if (n >= 0)
+                    {
+                        if (n < rel_distance)
+                        {
+                            it_ = begin + pos + n;
+                        }
+                        else
+                        {
+                            n -= rel_distance;
+                            it_ = begin + n % distance;
+                        }
+                    }
+                    else
+                    {
+                        n *= -1;
+                        if (n <= rel_distance)
+                        {
+                            it_ = begin + pos - n;
+                        }
+                        else
+                        {
+                            n -= rel_distance;
+                            it_ = begin + distance - n % distance;
+                            if (it_ == end)
+                            {
+                                it_ = begin;
+                            }
+                        }
+                    }
+                }
+
+                CONCEPT_REQUIRES(RandomAccessRange<Rng>())
+                difference_type distance_to(cursor const &that) const
+                {
+                    RANGES_ASSERT(that.rng_ == rng_);
+                    return that.pos_ - pos_;
+                }
+            };
+
+            template <bool IsConst>
+            struct sentinel
+            {
+                sentinel() = default;
+                constexpr bool equal(cursor<IsConst> const &) const
+                {
+                    return false;
+                }
+            };
```

Comment:
If you put a `done` method on the cursor, you don't need to define a sentinel at all.